### PR TITLE
DiskCache improvement

### PR DESCRIPTION
**Pull request by mickeylee25**
This PR includes:
1. Refactor DiskCache by conforming to interace `DiskCacheable`. This adds the ability to unit test.
2. Refactor `RemoteDiskCacheDataProviding` by injecting the DiskCachable.
3. Change the response type of `RequestDispatcher.dispatch(request: URLRequest)` from `Observable<Any>` to `Observable<Data>`. Thus, we can reduce the JSONSerialisation codes in the dispatcher and RemoteDiskCacheDataProviding.
4. Improve test coverage.
5. Resolve lint warings on APILayer/* except for Tests.

### CODE REVIEW

File: RealifeTech-CoreSDK/APILayer/Util/String+RequestIdentifier.swift [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+//  Created by Mickey Lee on 05/03/2021.
+//
+
+import Foundation
+import CryptoKit
+
+extension String {
+
+    var md5: String {
+        let digest = Insecure.MD5.hash(data: self.data(using: .utf8) ?? Data())
```
Comment (SoaringEarth) [inline RealifeTech-CoreSDK/APILayer/Util/String+RequestIdentifier.swift:14]: Should we look to improve this at some point?

File: RealifeTech-CoreSDK/APILayer/DiskCache/DiskCache.swift [hunk 2]
```diff
@@ -8,112 +8,169 @@
...
-                    guard let text = try? String(contentsOf: file.absoluteURL, encoding: .utf8) else { continue }
-                    shouldDelete = fileExpired(text: text, andAlsoOutdated: true)
+    public func deleteItem(with fileName: String) {
+        guard let fileUrl = try? getFileUrl(fileName) else { return }
+        try? fileManager.removeItem(at: fileUrl)
+    }
+
+    public func clearItems(deletionStrategy: DiskCacheDeletionStrategy, completion: (() -> Void)?) {
+        guard let files = try? getAllFilesUrl() else { return }
+        let fullExtension = deletionStrategy == .privateOnly ?
```
Comment (plspalding) [inline RealifeTech-CoreSDK/APILayer/DiskCache/DiskCache.swift:101]: What do you think of the following styling @mickeylee25 where we have the question mark and the : on the line with the value? I prefer this as the boolean expression is on the first line and the value associated with each statement are on their respective lines.

```
let fullExtension = deletionStrategy == .privateOnly 
    ? "\(Self.privateIndicator)\(Self.fileExtension)"
    : "\(Self.fileExtension)"
```

**Review by SoaringEarth — APPROVED** [review]

File: RealifeTech-CoreSDK/APILayer/Tests/DiskCacheTests.swift [hunk 3]
```diff
@@ -9,61 +9,163 @@
...
+        fileManager = MockFileManager()
+        clearCacheQueue = DispatchQueue(label: "DiskCacheTests Clear Cache Queue")
+        sut = DiskCache(fileManager: fileManager, clearCacheQueue: clearCacheQueue)
+    }
+
+    override func tearDownWithError() throws {
+        if fileManager.fileExists(atPath: expectedSavedPath.path) {
+            try fileManager.removeItem(at: expectedSavedPath)
+        }
+        if fileManager.fileExists(atPath: expectedSavedPath.path) {
```
Comment (plspalding) [inline RealifeTech-CoreSDK/APILayer/Tests/DiskCacheTests.swift:41]: This is duplicated from the lines above. Can it be removed?

File: RealifeTech-CoreSDK/APILayer/Util/String+RequestIdentifier.swift [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        if let allHTTPHeaderFields = allHTTPHeaderFields {
+            let safeHeaderFields = allHTTPHeaderFields
+                .filter { return $0.key != "Authorization" }
+                .sorted { $0.key < $1.key }
+                .map { "\($0.key)=\($0.value)" }
+                .joined(separator: ",")
+            string += safeHeaderFields.description
+        }
+        if let httpBody = httpBody, let bodyString = String(data: httpBody, encoding: .utf8) { string += bodyString }
+        return string.md5
```
Comment (plspalding) [inline RealifeTech-CoreSDK/APILayer/Util/String+RequestIdentifier.swift:36]: Am I right in thinking the md5 is being used to encrypt the body of the request we are making? Is this secure enough? I ask cos in the MD5 code it says `Insecure.MD5`.

File: RealifeTech-CoreSDK/APILayer/DiskCache/DiskCache.swift (hunk 2 above)
Comment (mickeylee25) [reply RealifeTech-CoreSDK/APILayer/DiskCache/DiskCache.swift:101]: I don't have objections to this. The format is used in another React project so I keep using here.

File: RealifeTech-CoreSDK/APILayer/Tests/DiskCacheTests.swift (hunk 3 above)
Comment (mickeylee25) [reply RealifeTech-CoreSDK/APILayer/Tests/DiskCacheTests.swift:41]: Good catch. This should be `expectedSavedPrivateObjectPath`

File: RealifeTech-CoreSDK/APILayer/Util/String+RequestIdentifier.swift (hunk 1 above)
Comment (mickeylee25) [reply RealifeTech-CoreSDK/APILayer/Util/String+RequestIdentifier.swift:14]: I haven't raised or seen the ticket for changing it to `sha256` but I can change it in this PR

File: RealifeTech-CoreSDK/APILayer/Util/String+RequestIdentifier.swift (hunk 4 above)
Comment (mickeylee25) [reply RealifeTech-CoreSDK/APILayer/Util/String+RequestIdentifier.swift:36]: Yes, you're right about the encryption. In terms of the security, I'm not 100% sure about it and as the above comment, I can change it to `sha256` in this PR.

File: RealifeTech-CoreSDK/APILayer/Util/String+RequestIdentifier.swift (hunk 1 above)
Comment (plspalding) [reply RealifeTech-CoreSDK/APILayer/Util/String+RequestIdentifier.swift:14]: I would raise a ticket @mickeylee25 . That way you can get this PR in.

**Review by plspalding — APPROVED** [review]

File: RealifeTech-CoreSDK/APILayer/Util/String+RequestIdentifier.swift (hunk 1 above)
Comment (mickeylee25) [reply RealifeTech-CoreSDK/APILayer/Util/String+RequestIdentifier.swift]: I just committed the changes of SHA256 to this PR.

File: RealifeTech-CoreSDK/APILayer/Util/String+RequestIdentifier.swift (hunk 4 above)
Comment (plspalding) [reply RealifeTech-CoreSDK/APILayer/Util/String+RequestIdentifier.swift]: I left comment @SoaringEarth comment around same issue.

File: RealifeTech-CoreSDK/APILayer/Util/String+RequestIdentifier.swift [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+//  Created by Mickey Lee on 05/03/2021.
+//
+
+import Foundation
+import CryptoKit
+
+extension String {
+
+    var sha256: String {
+        let digest = SHA256.hash(data: self.data(using: .utf8) ?? Data())
```
Comment (plspalding) [inline RealifeTech-CoreSDK/APILayer/Util/String+RequestIdentifier.swift:14]: Can the return statement be removed here as well as the reference to self by doing following:
```
var sha256: String {
    SHA256.hash(data: data(using: .utf8) ?? Data()).hexString
}
```

Comment (plspalding) [reply RealifeTech-CoreSDK/APILayer/Util/String+RequestIdentifier.swift:14]: If `data(using: .utf8)` fails, what does hashing `Data()` produce? Is it an empty string?

Comment (mickeylee25) [reply RealifeTech-CoreSDK/APILayer/Util/String+RequestIdentifier.swift:14]: I think it won't return an empty string if hashing Data(). Probably it'd be better to use a guard statement to ensure it returns an empty string like below. The data is non-optional in `SHA.hash(data: DataProtocol)` so it's unlikely to just use `SHA256.hash(data: data(using: .utf8) ?? Data()).hexString`
```
guard let data = self.data(using: .utf8) else { return "" }
let digest = SHA256.hash(data: data)
return digest.hexString
```

Comment (plspalding) [reply RealifeTech-CoreSDK/APILayer/Util/String+RequestIdentifier.swift:14]: The code I wrote was doing the same as your original code. I was just curious what it produced it `data(using: .utfi)` failed. Returning an empty string is probably better, as encrypting empty data doesn't seem useful.

Could use map:
```
data(using: .utf8).map { SHA.hash(data: $0).hexString } ?? ""
```

**Review by plspalding — APPROVED** [review]

**Pull request merged**